File: 17-11/subsets.py
class Solution:
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        paths = []
        self.gen_sub_recur(0,[],paths,nums)
        return paths

    def gen_sub_recur(self,index,sub,subs,nums):
        subs.append(sub)
        for i in range(index ,len(nums)):
            self.gen_sub_recur(i+1 , sub+[nums[i]],subs,nums)


Solu = Solution()
n = [1,2,3]
res = Solu.subsets(n)
print(res)

File: 17-11/Subsets.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
	vector<vector<int>> subsets(vector<int>& nums) {
		sort(nums.begin(), nums.end());
		vector<vector<int>> results;
		vector<int> path;
		gensub_recursive(nums, 0, path, results);
		return results;
	
	}

	void gensub_recursive(vector<int>& nums, int index, vector<int> &sub, vector<vector<int>> &subs) {
		subs.push_back(sub);
		for (int i = index; i < nums.size(); ++i) {
			sub.push_back(nums[i]);
			gensub_recursive(nums, i + 1, sub, subs);
			sub.pop_back();
		}
	}
};

int main()
{

}